internal/http/services/owncloud/ocdav/token.go

```diff
@@ -0,0 +1,171 @@
+			s.handleGetToken(w, r, tkn, c, false)
+		}
+	})
+}
+
+// TokenInfo contains information about the token
+type TokenInfo struct {
+	// for all callers
+	Token             string `xml:"token"`
+	LinkURL           string `xml:"linkurl"`
+	PasswordProtected bool   `xml:"passwordprotected"`
+
+	// if not password protected
+	StorageID string `xml:"storageid"`
+	OpaqueID  string `xml:"opaqueid"`
+	Path      string `xml:"path"`
+
+	// if native access
+	SpacePath  string `xml:"spacePath"`
+	SpaceAlias string `xml:"spaceAlias"`
+	SpaceURL   string `xml:"spaceURL"`
+}
+
+func (s *svc) handleGetToken(w http.ResponseWriter, r *http.Request, tkn string, c gateway.GatewayAPIClient, protected bool) {
+	ctx := r.Context()
+	log := appctx.GetLogger(ctx)
+
+	user, token, passwordProtected, err := getInfoForToken(tkn, r.URL.Query(), c)
+	if err != nil {
+		log.Error().Err(err).Msg("error stating token")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	t, err := buildTokenInfo(user, tkn, token, passwordProtected, c)
+	if err != nil {
+		log.Error().Err(err).Msg("error stating resource behind token")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	if protected {
+		if t.PasswordProtected == true {
+			log.Error().Msg("password protected private links are not supported")
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+		space, status, err := spacelookup.LookUpStorageSpaceByID(ctx, c, t.StorageID)
+		// add info only if user is able to stat
+		if err == nil && status.Code == rpc.Code_CODE_OK {
+			t.SpacePath = utils.ReadPlainFromOpaque(space.Opaque, "path")
+			t.SpaceAlias = utils.ReadPlainFromOpaque(space.Opaque, "spaceAlias")
+			t.SpaceURL = path.Join(t.SpaceAlias, t.OpaqueID, t.Path)
+		}
+
+	}
+
+	b, err := xml.Marshal(t)
+	if err != nil {
+		log.Error().Err(err).Msg("error marshaling xml")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	w.Write(b)
+	w.WriteHeader(http.StatusOK)
+	return
+}
+
+func buildTokenInfo(owner *user.User, tkn string, token string, passProtected bool, c gateway.GatewayAPIClient) (TokenInfo, error) {
+	t := TokenInfo{Token: tkn, LinkURL: "/s/" + tkn}
+	if passProtected {
+		t.PasswordProtected = true
+		return t, nil
+	}
+
+	ctx := ctxpkg.ContextSetToken(context.TODO(), token)
+	ctx = ctxpkg.ContextSetUser(ctx, owner)
+	ctx = metadata.AppendToOutgoingContext(ctx, ctxpkg.TokenHeader, token)
+
+	sRes, err := getTokenStatInfo(ctx, c, tkn)
+	if err != nil || sRes.Status.Code != rpc.Code_CODE_OK {
+		return t, fmt.Errorf("can't stat resource. %+v %s", sRes, err)
+	}
+
+	ls := &link.PublicShare{}
+	_ = json.Unmarshal(sRes.Info.Opaque.Map["link-share"].Value, ls)
+
+	t.StorageID = ls.ResourceId.GetStorageId()
+	t.OpaqueID = ls.ResourceId.GetOpaqueId()
+
+	return t, nil
+}
+
+func getInfoForToken(tkn string, q url.Values, c gateway.GatewayAPIClient) (owner *user.User, token string, passwordProtected bool, err error) {
+	ctx := context.Background()
+
+	sig := q.Get("signature")
```

Comment:
Will there ever be a signature in the URL? Also what about password authentication via basic auth?

Reply:
I just used same logic as with public links. Don't know if this is really needed.

Regarding second question: Even if there was a password authentication that still doesn't mean the user has access to the link. We use the link to determine the resourceID (and some other information) of the underlying resource in all cases.
I didn't see the point in trying to use the users access token, since he can just use the link token to get that information.

Reply:
> Even if there was a password authentication that still doesn't mean the user has access to the link.

I don't mean user authentication but a public link can be password protected.

Let's assume this file tree:
/
-+ Folder1 <- Shared with public link. Token: foobar42, password: secret
--+ Cat.jpg
-+ Folder2

Now if I would like to PROPFIND the share I would have to send:
```
curl -u public:secret -X PROPFIND https://localhost:9200/remote.php/dav/public-files/foobar42

The signature is also only present for password protected public links. It is used when clients want to download a file with a standard html link.
Then the link would look like this: `https://localhost:9200/remote.php/dav/public-files/foobar42/Cat.jpg?signature=<signature>&expiration=<timestamp>`

That is why I think you should rather handle the link password instead of the signature.

Reply:
Ah. Now I got it. Thanks for explanation :+1: 
Please check again, we are using `BaicAuth(...)` now.